grouper/permissions.py

```diff
@@ -0,0 +1,409 @@
+            continue
+        for name, permission_obj in all_permissions.iteritems():
+            if matches_glob(grantable[0], name):
+                result.append((permission_obj,
+                               grantable[1] if len(grantable) > 1 else '*', ))
+
+    return sorted(result, key=lambda x: x[0].name + x[1])
+
+
+def get_owners_by_grantable_permission(session):
+    """
+    Returns all known permission arguments with owners. This consolidates
+    permission grants supported by grouper itself as well as any grants
+    governed by plugins.
+
+    Args:
+        session(sqlalchemy.orm.session.Session): database session
+
+    Returns:
+        A map of permission to argument to owners of the form {permission:
+        {argument: [owner1, ...], }, } where 'owners' are models.Group objects.
+        And 'argument' can be '*' which means 'anything'.
+    """
+    all_permissions = {permission.name: permission for permission in Permission.get_all(session)}
+    all_groups = session.query(Group).filter(Group.enabled == True).all()
+
+    owners_by_arg_by_perm = defaultdict(lambda: defaultdict(list))
+    for group in all_groups:
+        grants = session.query(
+                Permission.name,
+                PermissionMap.argument,
+                PermissionMap.granted_on,
+                Group,
+        ).filter(
+                PermissionMap.group_id == Group.id,
+                Group.id == group.id,
+                Permission.id == PermissionMap.permission_id,
+                Permission.name == PERMISSION_GRANT,
+        ).all()
+
+        for perm, arg in filter_grantable_permissions(session, grants,
+                all_permissions=all_permissions):
+            owners_by_arg_by_perm[perm.name][arg].append(group)
+
+    # merge in plugin results
+    for plugin in get_plugins():
+        for (perm, arg), owners in plugin.get_perm_arg_to_owners(session).items():
+            owners_by_arg_by_perm[perm][arg] += owners
+
+    return owners_by_arg_by_perm
+
+
+def get_grantable_permissions(session):
+    """Returns all grantable permissions and their possible arguments. This
+    function attempts to reduce a permission's arguments to the most
+    permissive, i.e. if a wildcard argument exists, everything else is
+    discarded.
+
+    TODO: this only does a simplistic reduction and doesn't really understand
+    globbing.
+
+    Args:
+        session(sqlalchemy.orm.session.Session): database session
+
+    Returns:
+        A map of models.Permission object to a list of possible arguments, i.e.
+        {models.Permission: [arg1, arg2, ...], ...}
+    """
+    owners_by_arg_by_perm = get_owners_by_grantable_permission(session)
+    args_by_perm = defaultdict(list)
+    for permission, owners_by_arg in owners_by_arg_by_perm.items():
+        for argument in owners_by_arg:
+            args_by_perm[permission].append(argument)
+
+    def _reduce_args(args):
+        return "*" if len(args) > 1 and "*" in args else args
```

Comment:
Nit: I think this should be `["*"]`.  Because of Python's treatment of strings as sequences, it ends up working either way, but it confused me a bit when I realized this didn't return a list.